apps/aeutils/src/aeu_mtrees.erl

```diff
@@ -0,0 +1,93 @@
+%%%-------------------------------------------------------------------
+%%% @copyright (C) 2017, Aeternity Anstalt
+%%% @doc In-memory Merkle trees.
+%%%
+%%% The root hash depends on the order of the operations performed on
+%%% the tree.
+%%%
+%%% This implementation is a wrapper for `gb_merkle_trees` with
+%%% stricter checks on arguments, less ambiguous return values and
+%%% enabling better type specifications.
+%%%
+%%% @see gb_merkle_trees
+%%% @end
+%%%-------------------------------------------------------------------
+-module(aeu_mtrees).
+
+%% API
+-export([new/0,
+         get/2,
+         get_with_proof/2,
+         put/3,
+         root_hash/1,
+         verify_proof/4,
+         to_orddict/1]).
+
+-export_type([mtree/0,
+              mtree/2,
+              root_hash/0]).
+
+-define(HASH_BYTES, 32).
+-define(IS_KEY(K), is_binary(K)).
+-define(IS_VALUE(V), is_binary(V)).
+
+-type key() :: binary().
+-type value() :: binary().
+-type mtree() :: mtree(key(), value()).
+
+%% Enable specification of types of key and value for enabling code
+%% using this module to document types for readability.
+%% Both key and value must be binaries.
+-type mtree(_K, _V) :: gb_merkle_trees:tree().
+
+-type root_hash() :: <<_:(?HASH_BYTES*8)>>.
+
+%%%===================================================================
+%%% API
+%%%===================================================================
+
+-spec new() -> mtree().
+new() ->
+    {ok, gb_merkle_trees:empty()}.
```

Comment:
Why the {ok, _} wrapper?

Why don't you clean these out while you are doing such extensive refactoring anyway.

Reply:
Ok.
The only function that performs meaningful return value transformation is `root_hash`. For the other functions, the only value is stricter validation of args.

Reply:
... and I will follow more closely dict API.

Reply:
Actually I realized following gb_trees OTP module (as opposed to dict) makes more sense (this is what the underlying gb_merkle_trees module does), will do that. gb_trees too does not have unnecessary wrappers.
cc @gorillainduction